File: app/models/asset.py
"""Asset domain model."""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone


@dataclass
class Asset:
    """A tradable holding tracked by the portfolio."""

    id: int | None
    name: str
    symbol: str = ""
    quantity: float = 0.0
    avg_buy_price: float = 0.0
    current_price: float = 0.0
    notes: str = ""
    created_at: str = ""
    updated_at: str = ""

    @property
    def total_value(self) -> float:
        return self.quantity * self.current_price

    @property
    def cost_basis(self) -> float:
        return self.quantity * self.avg_buy_price

    @property
    def unrealized_pnl(self) -> float:
        return self.total_value - self.cost_basis

    @property
    def unrealized_pnl_pct(self) -> float:
        if self.cost_basis == 0:
            return 0.0
        return (self.unrealized_pnl / self.cost_basis) * 100.0

    @classmethod
    def from_row(cls, row) -> Asset:
        return cls(
            id=row["id"],
            name=row["name"],
            symbol=row["symbol"] or "",
            quantity=float(row["quantity"]),
            avg_buy_price=float(row["avg_buy_price"]),
            current_price=float(row["current_price"]),
            notes=row["notes"] or "",
            created_at=row["created_at"],
            updated_at=row["updated_at"],
        )

    def touch(self) -> None:
        self.updated_at = (
            datetime.now(timezone.utc)
            .replace(tzinfo=None)
            .isoformat(timespec="seconds")
        )
        if not self.created_at:
            self.created_at = self.updated_at

File: lib/domain/models/asset.dart
class Asset {
  Asset({
    this.id,
    required this.name,
    this.symbol = '',
    this.quantity = 0,
    this.avgBuyPrice = 0,
    this.currentPrice = 0,
    this.notes = '',
    this.createdAt = '',
    this.updatedAt = '',
  });

  int? id;
  String name;
  String symbol;
  double quantity;
  double avgBuyPrice;
  double currentPrice;
  String notes;
  String createdAt;
  String updatedAt;

  double get totalValue => quantity * currentPrice;
  double get costBasis => quantity * avgBuyPrice;
  double get unrealizedPnl => totalValue - costBasis;

  factory Asset.fromMap(Map<String, Object?> m) => Asset(
        id: m['id'] as int?,
        name: (m['name'] as String?) ?? '',
        symbol: (m['symbol'] as String?) ?? '',
        quantity: (m['quantity'] as num?)?.toDouble() ?? 0,
        avgBuyPrice: (m['avg_buy_price'] as num?)?.toDouble() ?? 0,
        currentPrice: (m['current_price'] as num?)?.toDouble() ?? 0,
        notes: (m['notes'] as String?) ?? '',
        createdAt: (m['created_at'] as String?) ?? '',
        updatedAt: (m['updated_at'] as String?) ?? '',
      );

  Map<String, Object?> toMap() => {
        if (id != null) 'id': id,
        'name': name,
        'symbol': symbol,
        'quantity': quantity,
        'avg_buy_price': avgBuyPrice,
        'current_price': currentPrice,
        'notes': notes,
        'created_at': createdAt,
        'updated_at': updatedAt,
      };
}
